src/server.c

```diff
@@ -677,6 +677,42 @@ int clientsCronResizeQueryBuffer(client *c) {
     return 0;
 }
 
+/* The client output buffer can be adjusted to better fit the memory requierments.
+ *
+ * the logic is:
+ * in case the last observed peak size of the buffer equals the buffer size - we double the size
+ * in case the last observed peak size of the buffer equals less than half the buffer size - we shrink by half.
+ * in any case we will reset the peak value each cron run.
```

Comment:
redis has a mechanism that attempts to scan through all clients once a second, but it also handles at least 10 in each cron, so if there are not a lot of clients, they can be scanned faster.
this means that resetting the peak value on each cron, can behave differently when there are a lot of clients vs just a few.
i think it makes sense to add a mechanism to avoid doing this job (and reset the peak) too often.

Reply:
i'm sorry, i suppose it wasn't clear what i commented on (i selected one line, but the context in GH shows more).
i was specifically referring to:
```
in any case we will reset the peak value each cron run
```

Reply:
@oranagra I agree that we can soften the peek value reset in some way.
I thought that keeping a new per-client indicator for how many cron iterations visited that client can help in more future cases and can help avoid reseting the peek too often in this case..
alternatively I can just keep the last servertime visiting this client in cron...
what do you think? would that answer your raised concerns?

Reply:
not sure i understand how counting cron iterations will help.
i think we need to remember the peak reset time, and avoid checking these thresholds (and resetting the peak) too soon.
maybe it could be some generic clientsCron mechanism that can serve other per-client cron tasks later.
i.e some clients cron are better be done as soon as possible, but others should not be done too soon.
so there could be a block in clientsCron that's executed just once a second, saving the last time it run in the clients struct.

Reply:
@oranagra what I meant is that keeping a per-client track of how many times it was handled by the clients cron can be used in different scenarios to make operations based on estimated time without having to keep a dedicated variables for each.
For example in this case I can reset the peak every (client_cron_ticks % (100/server_hz))
But I think we can currently keep it simple and just make sure we reset the peak value each 5 seconds.
Does that sound better?

Reply:
@ranshid since in certain cases a client can be hit every cron tick, and in others just once a second, i don't think incrementing a counter every time we hit that client is good.
i.e. i assume you mean some mechanism like run_with_period, but that can work only if you know you're incrementing as a roughly constant pace.
maybe i'm misunderstanding you, so writing some pseudo code, or just posting the implementation could help.

we could instead save the value of `server.cronloops` on the last time we hit that client, but i don't even see how this would help to serve multiple purposes, since we have to reset that variable when we perform the action, so if we have multiple actions with different timing requirements, we'll need multiple variables.

my suggestion is able to serve multiple purposes with one variable, just because it defines that all these purposes need a one second intervals.
i.e. the clientsCron is divided into two groups, one runs on every call, and the other is throttled to once a second.

Reply:
@oranagra I agree with all you are saying. correct me if I am wrong but the current clients cron is exactly an attempt to operate each client every second so . I believe that most of these operations do not really need to be operated exactly once every second but rather as an estimation to prevent keeping the server busy. My main point was trying to make peak resets happen less often which is why I suggested to keep track of cron ticks and perform the operation once every x ticks. I understand that this will cause some difference between scenarios of few and many clients but I am not sure that is a problem. in case of few clients the peak will be checked more often but also the transmit operations, so in case the client is taking on more work I would also expect the peak to reflect the correct reply buffer usage.  In order to reduce the risk I currently set the peak reset period to be once every 5 seconds which is fine as we are mainly worried about un needed shrinks. I just do not want this to become a very complicated solution and keep the clients cron simple and fast.

Reply:
> I believe that most of these operations do not really need to be operated exactly once every second but rather as an estimation to prevent keeping the server busy

that's right, but the mechanism is to make sure they're handled **at least** once a second, and it could be much faster.

> in case of few clients the peak will be checked more often but also the transmit operations

not sure why you say that, i don't see the correlation between client count and the amount of work / data they induce.

Reply:
@oranagra I understand. actually I meant exactly something like run_with_period (just was not familiar with it)
in any case my last update makes each client reset the peak every 5 seconds, so it should resolve the few clients issue.
Unless I missed something :)